ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -622,6 +622,12 @@ static const upb_pbdecodermethod *msgdef_decodermethod(Descriptor* desc) {
   return desc->fill_method;
 }
 
+// Callback invoked by upb if any error occurs during parsing or serialization.
+static bool env_error_func(void* ud, const upb_status* status) {
+  const char* message = (const char*)ud;
+  rb_raise(rb_eRuntimeError, message, upb_status_errmsg(status));
```

Comment:
Makes sense. I think that this should be compliant with the contract you describe -- `rb_raise` will unwind the whole C stack, all the way back up into Ruby.